### PR TITLE
feat: Create new rule `avoid_collection_methods_with_unrelated_types`

### LINKED ISSUE #701: [New rule] `avoid_collection_methods_with_unrelated_types` (Willing to PR)
**Please describe what the rule should do:**
Please see https://github.com/dart-lang/linter/issues/1307

**If your rule is inspired by other please provide link to it:**
https://github.com/dart-lang/linter/issues/1307
https://stackoverflow.com/questions/68251477/dart-flutter-linter-rule-the-type-to-index-a-map-should-be-the-key-type-of-map

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

```dart
https://github.com/dart-lang/linter/issues/1307
```

**Are you willing to submit a pull request to implement this rule?**
Yes

### PR DESCRIPTION
**Pull request by fzyzcjy**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[x] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### What changes did you make? (Give an overview)

Close #701

### CODE REVIEW

**Comment by fzyzcjy** [comment]
Tests have passed locally :)

File: lib/src/analyzers/lint_analyzer/models/issue.dart [hunk 1]
```diff
@@ -43,4 +43,14 @@ class Issue {
     this.verboseMessage,
     this.suggestion,
   });
+
+  @override
+  String toString() => 'Issue{'
```
Comment (incendial) [inline lib/src/analyzers/lint_analyzer/models/issue.dart:48]: Could you point out how this change is used?

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_collection_methods_with_unrelated_types/visitor.dart [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        final mapType = _getMapType(node.target?.staticType);
+        final argType = node.argumentList.arguments.singleOrNull?.staticType;
+        if(mapType != null && argType != null && argType.asInstanceOf(mapType.keyElement) == null) {
+          _expressions.add(node);
+        }
+        break;
+      case 'containsValue':
+        final mapType = _getMapType(node.target?.staticType);
+        final argType = node.argumentList.arguments.singleOrNull?.staticType;
+        if(mapType != null && argType != null && argType.asInstanceOf(mapType.valueElement) == null) {
```
Comment (incendial) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_collection_methods_with_unrelated_types/visitor.dart:47]: Also, please run dart format on the codebase

**Comment by fzyzcjy** [comment]
> I guess, the main questions is about supporting not only maps, but sets and lists too. What do you think?

Following the original issue, I may also add `Iterable.contains`.

> Also, was it hard to add a rule? Maybe you have faced some problems that we should address on order to make the contribution process easier?

I just copy-and-paste-and-modify from sibling rules ;) It is ok imho

File: lib/src/analyzers/lint_analyzer/models/issue.dart (hunk 1 above)
Comment (fzyzcjy) [reply lib/src/analyzers/lint_analyzer/models/issue.dart:48]: Logging for debugging what is that issue. I can revert if needed.

Comment (fzyzcjy) [reply lib/src/analyzers/lint_analyzer/models/issue.dart]: reverted

File: lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_collection_methods_with_unrelated_types/visitor.dart [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+          _expressions.add(node);
+        }
```
[the remark this answers is not in the document]
Comment (fzyzcjy) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/avoid_collection_methods_with_unrelated_types/visitor.dart]: done

File: website/docs/rules/common/avoid-collection-methods-with-unrelated-types.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+```dart
+final map = Map<int, String>();
+map["str"] = "value";
+```
+
+Good:
+
+```dart
+final map = Map<int, String>();
+map[42] = "value";
```
[the remark this answers is not in the document]
Comment (fzyzcjy) [reply website/docs/rules/common/avoid-collection-methods-with-unrelated-types.md:47]: done

File: website/docs/rules/common/avoid-collection-methods-with-unrelated-types.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+Warning
+
+## Description {#description}
+
+Avoid using collection methods with unrelated types, such as accessing a map of integers using a string key.
+
+This lint has been requested for a long time: Follow [this link](https://github.com/dart-lang/linter/issues/1307) to see the details.
+
+Related: Dart's built-in `list_remove_unrelated_type` and `iterable_contains_unrelated_type`.
```
[the remark this answers is not in the document]
Comment (fzyzcjy) [reply website/docs/rules/common/avoid-collection-methods-with-unrelated-types.md:17]: done

**Comment by fzyzcjy** [comment]
@incendial Done and ready for review (again(

**Comment by incendial** [comment]
@fzyzcjy looks amazing, thanks for addressing all the questions!

 > I just copy-and-paste-and-modify from sibling rules ;) It is ok imho
 
Thanks for answering, we don't have many people contributing new rules, so I was wondering, if it's something wrong with the flow.

**Comment by incendial** [comment]
The only thing missing is the entry in changelog under `Unreleased` section, like `feat: add static code diagnostic avoid-collection-methods-with-unrelated-types`. Could you add it?

Now we'll run the rule on different public codebases in order to avoid possible edge-cases.

**Review by incendial — APPROVED** [review]

**Comment by fzyzcjy** [comment]
You are welcome! Done.

**Review by dkrutskikh — APPROVED** [review]

**Comment by dkrutskikh** [comment]
@fzyzcjy thank you for contribute

**Comment by fzyzcjy** [comment]
You are welcome!

**Pull request merged**